### ISSUE TITLE
defined config overwritten by config further up directory tree

### ISSUE THREAD

**Issue by jsnklln** [jcrocholl/pep8#368]
In the read_config function first the "user" config file is read then the directory tree is searched for "local" config files.  If any config files exist within the directory tree they overwrite what was defined in the file specified by using --config.   I've made a fork and plan on issuing a pull request soon.  BTW I have no idea what I'm doing.

**Comment by IanLee1521** [comment]
Closed with commit 19d9a82

**Comment by sigmavirus24** [comment]
It seems like the fix for this issue essentially broke people relying on config files in `$HOME/.config/`. See also: https://gitlab.com/pycqa/flake8/issues/33

**Comment by jsnklln** [comment]
I've got to leave for the weekend but I wanted to include what I've found so far:

    flake8 --version
    2.3.0 (pep8: 1.6.1, mccabe: 0.3, pyflakes: 0.8.1) CPython 2.7.9 on Darwin

My python code:
`x = "fsafdsafdsafdsafdsafdsafdsafdsafdsafdsafdsajfkdsajkfdsajkfdsjaklfdjsaklfjdsaklfjdkslajfkldsajklfdsajklfjdsaklfdsa"`

My config:

    [flake8]
    max-line-length=100

Run it:

    flake8 code.py
    code.py:1:101: E501 line too long (119 > 100 characters)

That's goodish news.

In the process I ran across the fact that the .config/pep8 file is read when flake8 is run but the [flake8] section is used.  When pep8 is run the .config/pep8 file is read but the [pep8] section is used.  I haven't looked into why flake8 isn't pulling in the right file.

**Comment by sigmavirus24** [comment]
@IanLee1521 should I file a separate issue for the regression or should we reopen this to cover it?

**Comment by IanLee1521** [comment]
No, I think this was a regression introduced with some cleanup I did. See #380 for more details.

I've tested against flake8 2.3.0 and commit 3bbaa23 and this seems to be resolved.

```
$ cat bar.py
x = "fsafdsafdsafdsafdsafdsafdsafdsafdsafdsafdsajfkdsajkfdsajkfdsjaklfdjsaklfjdsaklfjdkslajfkldsajklfdsajklfjdsaklfdsa"

$ cat ~/.config/flake8
[flake8]
max-line-length=100

$ flake8 --version
2.3.0 (pep8: 1.6.1, mccabe: 0.3, pyflakes: 0.8.1) CPython 2.7.6 on Linux

$ flake8 bar.py
bar.py:1:80: E501 line too long (119 > 79 characters)

$ flake8 --version
2.3.0 (pep8: 1.6.1, mccabe: 0.3, pyflakes: 0.8.1) CPython 2.7.6 on Linux

$ flake8 bar.py
bar.py:1:101: E501 line too long (119 > 100 characters)
```

**Comment by IanLee1521** [comment]
I'm going to reopen this issue. I've reverted the majority of this fix in commit 435d1cb as it had some pretty serious repurcussions for flake8 users. I intend to circular back around and propose an improved solution to this that works for both pep8 and flake8, however, I've unfortunately not gotten as much time as I'd hoped to devote to that work.

@jsnklln -- In response to:

> In the process I ran across the fact that the .config/pep8 file is read when flake8 is run but the [flake8] section is used. When pep8 is run the .config/pep8 file is read but the [pep8] section is used. I haven't looked into why flake8 isn't pulling in the right file.

I believe you're looking for https://gitlab.com/pycqa/flake8/blob/master/flake8/engine.py#L51 which returns a parser that looks for the `[flake8]` rather than `[pep8]` section in the config file.
